vmss-prototype/vmss-prototype

```diff
@@ -851,13 +872,148 @@ def vmss_prototype_update(sub_args):
+    sig_name = get_vmss_sig(resource_group)
+
+    if sub_args.target_vmss:
+        vmss_names = sorted(set([name for names in sub_args.target_vmss for name in names]))
+    else:
+        # Get all of the VMSSes in this cluster
+        vmss_names = get_vmss_set()
+
+    # Get the set of unique annotations that would trigger a skip
+    node_skip_annotations = set([name for names in sub_args.node_skip_annotation for name in names]) if sub_args.node_skip_annotation else set()
+
+    # Get the total node data such that we can make some judgements as to fitness
+    all_nodes_json, _, _ = run(['kubectl', 'get', 'nodes',
+                                '--output', 'json'
+                                ], retries=3)
+
+    nodes = json.loads(all_nodes_json)['items']
+
+    for vmss in vmss_names:
+        # The VMSS specific image definition name
+        image_definition = get_sig_image_def(vmss)
+
+        versions = get_vmss_images(subscription, resource_group, sig_name, image_definition)
+        # If we don't have one, any version is newer than this :-)
+        latest_image = '0000.00.00'
+        if versions:
+            for version in versions:
+                if version['name'] > latest_image:
+                    latest_image = version['name']
+        logging.debug('Latest image for VMSS {0} is {1}'.format(vmss, latest_image))
+
+        potential_nodes = []
+        for node in nodes:
+            metadata = node.get('metadata', {})
+            # If not a node in target VMSS, continue...
+            node_name = metadata.get('name', 'unknown')
+            if vmss not in node_name:
+                logging.debug('SKIP: Node {0} not part of vmss {1}'.format(node_name, vmss))
+                continue
+
+            if node_name == sub_args.running_on_node:
+                logging.debug('SKIP: Node {0} is the same as the node we are running on')
+                continue
+
+            annotations = metadata.get('annotations', {})
+            skip = False
+            for skip_annotation in node_skip_annotations:
+                if skip_annotation in annotations:
+                    logging.debug('SKIP: Node {0} is annotated as needing to be skipped: {1}'.format(node_name, skip_annotation))
+                    skip = True
+                    break
+            if skip:
+                continue
+
+            if sub_args.pending_reboot_annotation in annotations:
+                logging.debug('SKIP: Node {0} is annotated as pending reboot: {1}'.format(node_name, sub_args.pending_reboot_annotation))
+                continue
+
+            last_patch = annotations.get(sub_args.last_patch_annotation, None)
+            if not last_patch:
+                logging.debug('SKIP: Node {0} does not have a last patch annotation: {1}'.format(node_name, sub_args.last_patch_annotation))
+                continue
+
+            last_patch_date_match = re.search(r'\d\d\d\d-\d\d-\d\d', last_patch)
```

Comment:
I would prefer if we standardize to RFC3339 for our time type to accommodate more fine-grained timestamps. If we want to be opinionated about only allowing a day's worth of resolution, we can easily convert RFC3339 to a lower-resolution format.

Reply:
I was doing this such that our (Skyman) current annotations could work as our annotations include additional information plus the timezone offset syntax (such as "reason@2021-01-09T09:32:37+00:00")

Reply:
Note that this whole process is just to produce a "version" string that can map with the shared image gallery version.

Reply:
I see that the regex match pattern is compatible w/ RFC3339, nothing to do here